bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryMetastoreClientImpl.java

```diff
@@ -0,0 +1,579 @@
+import org.apache.iceberg.exceptions.ServiceFailureException;
+import org.apache.iceberg.exceptions.ServiceUnavailableException;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+/** A client of Google Bigquery Metastore functions over the BigQuery service. */
+public final class BigQueryMetastoreClientImpl implements BigQueryMetastoreClient {
+
+  private final Bigquery client;
+  private final BigQueryOptions bigqueryOptions;
+
+  public static final ExceptionHandler.Interceptor EXCEPTION_HANDLER_INTERCEPTOR =
+      new ExceptionHandler.Interceptor() {
+
+        @Override
+        public RetryResult afterEval(Exception exception, RetryResult retryResult) {
+          return ExceptionHandler.Interceptor.RetryResult.CONTINUE_EVALUATION;
+        }
+
+        @Override
+        public RetryResult beforeEval(Exception exception) {
+          if (exception instanceof BaseServiceException) {
+            boolean retriable = ((BaseServiceException) exception).isRetryable();
+            return retriable
+                ? ExceptionHandler.Interceptor.RetryResult.RETRY
+                : ExceptionHandler.Interceptor.RetryResult.CONTINUE_EVALUATION;
+          }
+
+          return ExceptionHandler.Interceptor.RetryResult.CONTINUE_EVALUATION;
+        }
+      };
+
+  // Retry config with error messages and regex for rate limit exceeded errors.
+  private static final BigQueryRetryConfig DEFAULT_RETRY_CONFIG =
+      BigQueryRetryConfig.newBuilder()
+          .retryOnMessage(BigQueryErrorMessages.RATE_LIMIT_EXCEEDED_MSG)
+          .retryOnMessage(BigQueryErrorMessages.JOB_RATE_LIMIT_EXCEEDED_MSG)
+          .retryOnRegEx(BigQueryErrorMessages.RetryRegExPatterns.RATE_LIMIT_EXCEEDED_REGEX)
+          .build();
+
+  public static final ExceptionHandler BIGQUERY_EXCEPTION_HANDLER =
+      ExceptionHandler.newBuilder()
+          .abortOn(RuntimeException.class)
+          // Retry on connection failures due to transient network issues.
+          .retryOn(java.net.ConnectException.class)
+          // Retry to recover from temporary DNS resolution failures.
+          .retryOn(java.net.UnknownHostException.class)
+          .addInterceptors(EXCEPTION_HANDLER_INTERCEPTOR)
+          .build();
+
+  /** Constructs a client of the Google BigQuery service. */
+  public BigQueryMetastoreClientImpl(BigQueryOptions options)
+      throws IOException, GeneralSecurityException {
+    // Initialize client that will be used to send requests. This client only needs to be created
+    // once, and can be reused for multiple requests
+    HttpCredentialsAdapter httpCredentialsAdapter =
+        new HttpCredentialsAdapter(
+            GoogleCredentials.getApplicationDefault().createScoped(BigqueryScopes.all()));
```

Comment:
This credential relies on environment variables by default, right? How can we connect to multiple BigQuery projects  from a single machine? Why not add a dedicated property with `GoogleCredentials#fromStream` or something?

Reply:
Thank you @ebyhr Good catch! If it is not very important for you. I would like to address that support after merging this. I already received so much comments on this PR. I want to merge this first. This PR will cover most of use cases and also provide main basic functionality for GCP users.

Reply:
https://github.com/apache/iceberg/issues/13036